.ebignore includes symlinked files in ignored directories

<!-- Make sure we don't have an existing Issue that reports the bug you are seeing (both open and closed). 
If you do find an existing Issue, re-open or add a comment to that Issue instead of creating a new one. If your issue is not specific to aws-elastic-beanstalk-cli try posting it to the AWS Elastic Beanstalk forum(https://forums.aws.amazon.com/forum.jspa?forumID=86).-->

### Description

Having specified `.ebignore` to ignore a whole directory, if that directory contains a symlink, that symlink is not ignored and included in the source bundle that eb cli creates.

### Steps to reproduce

With an eb application and environment set up - where the project contains a directory called `ignore` and, within that, one standard directory `dir` and one symlinked directory `symlink`; and the project's `.ebignore` file contains an entry to ignore that whole directory from the created artifact - run `eb deploy env-name`.

`.ebignore` file:
```
/bin/ignore/
```
### Observed result

The artifact created and uploaded to S3 will contain symlinked directory `ignore/symlink`, but _not_ `ignore/dir`.

Structure of example project `eb-bug`:

![image](https://user-images.githubusercontent.com/6698306/111899546-36a2e400-8a25-11eb-84a3-5822a01a04af.png)

Actual (partial) output from `eb deploy env-name --debug` shows no entry at all for `./bin/ignore/symlink`:

```
...
2021-03-21 08:28:27,522 (INFO) eb : Found .ebignore, using system zip.
2021-03-21 08:28:27,522 (DEBUG) ebcli.core.fileoperations : Project root found at: /path/2/eb-bug
2021-03-21 08:28:27,522 (INFO) eb : Zipping up folder at location: /path/2/eb-bug
2021-03-21 08:28:27,523 (INFO) eb :  +adding: .//
2021-03-21 08:28:27,523 (INFO) eb :   +adding: ./.envrc
2021-03-21 08:28:27,524 (INFO) eb :   +adding: ./Issue.md
2021-03-21 08:28:27,524 (INFO) eb :   +adding: ./.gitignore
2021-03-21 08:28:27,524 (INFO) eb :   -skipping: ./.ebignore
2021-03-21 08:28:27,524 (INFO) eb :  +adding: ./bin/
2021-03-21 08:28:27,525 (INFO) eb :   +adding: ./bin/cli
2021-03-21 08:28:27,525 (INFO) eb :   -skipping: ./bin/ignore/dir/file
2021-03-21 08:28:27,525 (INFO) eb :   -skipping: ./.elasticbeanstalk
2021-03-21 08:28:27,525 (INFO) eb :   -skipping: ./.elasticbeanstalk/app_versions
2021-03-21 08:28:27,525 (INFO) eb :  +adding: ./.git/
...
```

### Expected result

The artifact created and uploaded to S3 should not contain directory `ignore`.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: macos 11.2.3
2. EBCLI version: EB CLI 3.19.3 (Python 3.9.2)

Any updates on this issue? 
IMO It should be a **high** priority issue. If one folder is ignored, it should be ignored entirely, included its symlinks to folders.

Without this fix, Elastic Beanstalk app.zip can end with a bunch of symlinks that point to undefined locations. For example, I can have this error on startup:

`2022/04/21 09:11:08.715481 [ERROR] An error occurred during execution of command [app-deploy] - [StageApplication]. Stop running the command. Error: chown /var/app/staging/venv/dev/lib/python3.8/collections: no such file or directory `

Because 'collections' is a symlink to a folder that do not exist in AL2.

All workarounds I am studying are quite painful… so it would be great to fix it.

--- a/ebcli/core/fileoperations.py
+++ b/ebcli/core/fileoperations.py
@@ -18,13 +18,13 @@ import os
 import shutil
 import stat
 import sys
-import warnings
-import yaml
 import zipfile
+import yaml
+import warnings
+from pathspec import PathSpec
 
 from cement.utils.misc import minimal_logger
 from ebcli.objects.buildconfiguration import BuildConfiguration
-from pathspec import PathSpec
 from six import StringIO
 from yaml import safe_load, safe_dump
 from yaml.parser import ParserError
@@ -431,6 +431,21 @@ def _zipdir(path, zipf, ignore_list=None):
         if '.elasticbeanstalk' in root:
             io.log_info('  -skipping: {}'.format(root))
             continue
+        for d in dirs:
+            cur_dir = os.path.join(root, d)
+            if os.path.islink(cur_dir):
+                # os.walk categorize symlinks-to-directories as dirs
+                # and we want to include symlinks in the zip
+                if cur_dir in ignore_list:
+                    io.log_info(' -skipping: {}'.format(cur_dir))
+                else:
+                    zipInfo = zipfile.ZipInfo()
+                    zipInfo.filename = os.path.join(root, d)
+
+                    # 2716663808L is the "magic code" for symlinks
+                    zipInfo.external_attr = 2716663808 if sys.version_info > (3,) else long(2716663808)
+
+                    zipf.writestr(zipInfo, os.readlink(cur_dir))
         for f in files:
             cur_file = os.path.join(root, f)
 

--- a/tests/unit/core/test_fileoperations.py
+++ b/tests/unit/core/test_fileoperations.py
@@ -717,16 +717,22 @@ ccc""",
     def test_zip_up_project(self, _validate_file_for_archive_mock):
         _validate_file_for_archive_mock.side_effect = lambda f: not f.endswith('.sock')
         shutil.rmtree('home', ignore_errors=True)
-        os.mkdir('ignore-me')
-        os.mkdir('linkme')
+        os.mkdir('ignoredir')
+        os.mkdir('linkdir')
         os.mkdir('src')
         os.mkdir(os.path.join('src', 'lib'))
-        open(os.path.join('src', 'lib', 'app.py'), 'w').write('import os')
-        open(os.path.join('src', 'lib', 'app.py~'), 'w').write('import os')
-        open(os.path.join('src', 'lib', 'ignore-this-file.py'), 'w').write('import os')
-        open(os.path.join('src', 'lib', 'test.sock'), 'w').write('mock socket file')
-        open(os.path.join('ignore-me', 'text.txt'), 'w').write('import os')
-        open(os.path.join('linkme', 'foobar.txt'), 'w').write('import os linkme')
+        with open(os.path.join('src', 'lib', 'app.py'), 'w') as f:
+            f.write('import os')
+        with open(os.path.join('src', 'lib', 'app.py~'), 'w') as f:
+            f.write('import os')
+        with open(os.path.join('src', 'lib', 'ignore-this-file.py'), 'w') as f:
+            f.write('import os')
+        with open(os.path.join('src', 'lib', 'test.sock'), 'w') as f:
+            f.write('mock socket file')
+        with open(os.path.join('ignoredir', 'ignoredfile.txt'), 'w') as f:
+            f.write('import os')
+        with open(os.path.join('linkdir', 'foobar.txt'), 'w') as f:
+            f.write('import os linkdir/foobar.txt')
 
         os.symlink(
             os.path.abspath(os.path.join('src', 'lib', 'app.py')),
@@ -742,8 +748,8 @@ ccc""",
                 target_is_directory=True
             )
             os.symlink(
-                os.path.abspath(os.path.join('linkme')),
-                os.path.abspath(os.path.join('ignore-me','linkme')),
+                os.path.abspath(os.path.join('linkdir')),
+                os.path.abspath(os.path.join('ignoredir', 'symlink-to-linkdir')),
                 target_is_directory=True
             )
         else:
@@ -751,12 +757,27 @@ ccc""",
                 os.path.join('src', 'lib', 'api'),
                 os.path.join('src', 'lib', 'api-copy')
             )
+            os.symlink(
+                os.path.abspath(os.path.join('linkdir')),
+                os.path.abspath(os.path.join('ignoredir', 'symlink-to-linkdir'))
+            )
 
-        open(os.path.join('src', 'lib', 'api', 'api.py'), 'w').write('import unittest')
+        with open(os.path.join('src', 'lib', 'api', 'api.py'), 'w') as f:
+            f.write('import unittest')
 
         fileoperations.zip_up_project(
             'app.zip',
-            ignore_list=[os.path.join('src', 'lib', 'ignore-this-file.py'), os.path.join('ignore-me/'), os.path.join('src', 'lib', 'api-copy')]
+            ignore_list=[
+                os.path.join('src', 'lib', 'ignore-this-file.py'),
+
+                # When a directory is specified in .ebignore, the directory itself and all of its
+                # contents (files and subdirectories) are automatically added to the ignore_list recursively
+                os.path.join('ignoredir'),
+                os.path.join('ignoredir', 'ignoredfile.txt'),
+                os.path.join('ignoredir', 'symlink-to-linkdir'),
+
+                os.path.join('src', 'lib', 'api-copy')
+            ]
         )
 
         os.mkdir('tmp')
@@ -769,7 +790,7 @@ ccc""",
         self.assertFalse(os.path.exists(os.path.join('tmp', 'src', 'lib', 'app.py~')))
         self.assertFalse(os.path.exists(os.path.join('tmp', 'src', 'lib', 'ignore-this-file.py')))
         self.assertFalse(os.path.exists(os.path.join('tmp', 'src', 'lib', 'test.sock')))
-        self.assertFalse(os.path.exists(os.path.join('tmp','ignore-me', 'link-me')))
+        self.assertFalse(os.path.exists(os.path.join('tmp', 'ignoredir', 'symlink-to-linkdir')))
 
     def test_delete_app_versions(self):
         os.mkdir(os.path.join('.elasticbeanstalk', 'app_versions'))

